### PR TITLE
Fixed mercenary loyalty issue

### PR DESCRIPTION
**Pull request by IsaacLic**
# Description
This fixes the issue where mercenaries are hostile to the player. It does so by maintaining the MercItem that the mercenary is tied to when it's converted to a FarShip and back. Now the mercenaries will remain loyal forever.

# Testing
Hire some mercenaries. They will stay loyal, even if they are far enough away that they are converted into FarShips.

### CODE REVIEW

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

File: engine/src/main/java/org/destinationsol/game/ship/FarShip.java [hunk 1]
```diff
@@ -93,8 +90,12 @@ public boolean shouldBeRemoved(SolGame game) {
 
     @Override
     public SolShip toObject(SolGame game) {
-        return game.getShipBuilder().build(game, position, velocity, angle, rotationSpeed, pilot, container, hullConfig, life, gun1,
+        SolShip ship = game.getShipBuilder().build(game, position, velocity, angle, rotationSpeed, pilot, container, hullConfig, life, gun1,
                 gun2, removeController, engine, repairer, money, tradeContainer, shield, armor);
+        if(isMerc){
```
[the remark this answers is not in the document]
Comment (Adrijaned) [reply engine/src/main/java/org/destinationsol/game/ship/FarShip.java]: Or just press Ctrl+Alt+L (or Shift+Alt+L) every once in a while and IDEA should fix these itself

File: engine/src/main/java/org/destinationsol/game/ship/FarShip.java [hunk 2]
```diff
@@ -189,4 +190,24 @@ public float getMoney() {
     public ItemContainer getIc() {
         return container;
     }
+
+    /**
+     * Each FarShip could be a mercenary. Each mercenary is definitely a FarShip .
```
Comment (Adrijaned) [inline engine/src/main/java/org/destinationsol/game/ship/FarShip.java]: This really should not be in the method javadoc, but rather class javadoc. Also, javadoc should start with one sentence summing the rest of the javadoc, ended by dot and a newline

Comment (Adrijaned) [inline engine/src/main/java/org/destinationsol/game/ship/FarShip.java]: This really should not be a part of method-level javadoc, but rather a class-level javadoc. Also, javadoc should start with a single sentence summing the shole rest of the javadoc, ended by a dot and separated from the rest by a blank line.

File: engine/src/main/java/org/destinationsol/game/ship/FarShip.java [hunk 3]
```diff
@@ -189,4 +190,24 @@ public float getMoney() {
...
+     * This method sets the associated MercItem.
+     * @param mercItem The {@link MercItem} that this FarShip is associated with. Can be null.
+     */
+    public void setMerc(MercItem mercItem) {
+        this.mercItem = mercItem;
+        isMerc = true;
+    }
+
+    /**
+     * Each FarShip could be a mercenary. Each mercenary is definitely a FarShip .
```
Comment (Adrijaned) [inline engine/src/main/java/org/destinationsol/game/ship/FarShip.java]: See above

File: engine/src/main/java/org/destinationsol/game/ship/FarShip.java [hunk 4]
```diff
@@ -189,4 +190,24 @@ public float getMoney() {
     public ItemContainer getIc() {
         return container;
     }
+
+    /**
+     * Each FarShip could be a mercenary. Each mercenary is definitely a FarShip .
+     * This method sets the associated MercItem.
```
Comment (Adrijaned) [inline engine/src/main/java/org/destinationsol/game/ship/FarShip.java]: Not really needed considering the signature of the method. Better to rephrase as `Sets this ship's MercItem`

File: engine/src/main/java/org/destinationsol/game/ship/FarShip.java [hunk 5]
```diff
@@ -189,4 +190,24 @@ public float getMoney() {
...
+     * @param mercItem The {@link MercItem} that this FarShip is associated with. Can be null.
```
Comment (Adrijaned) [inline engine/src/main/java/org/destinationsol/game/ship/FarShip.java]: When something can be null, use `Optional`. Or at least `@Nullable` annotation

File: engine/src/main/java/org/destinationsol/game/ship/SolShip.java [hunk 6]
```diff
@@ -238,13 +243,15 @@ public void update(SolGame game) {
...
-        if (FactionInfo.getDisposition().get(factionID) < -5) {
+
+        if (!isMerc && FactionInfo.getDisposition().get(factionID) < -5) {
             getPilot().stringToFaction("ehar");
         }
         else {
             getPilot().stringToFaction("laani");
         }
 
+
```
Comment (Adrijaned) [inline engine/src/main/java/org/destinationsol/game/ship/SolShip.java]: Unnecesary newline

File: engine/src/main/java/org/destinationsol/game/ship/FarShip.java (hunk 2 above)
Comment (Adrijaned) [reply engine/src/main/java/org/destinationsol/game/ship/FarShip.java]: Wait. Now I got to the end of the changes and... 
1) Fine, I'll let the javadoc formatting slip for GhostInTheMachine
2) Your javadoc doesn't really make sense, and the SolShip's should be changed too

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by jellysnake** [comment]
If the presence of a `mercItem` always indicates that they are, well, a merc then a seperate Boolean field is not strictly needed. It's also not a harm however and so it can be left in.

What I do want to see is a public `isMerc()` method that checks either the bool or the presence of a `mercItem`.

**Comment by GooeyHub** [comment]
Hooray Jenkins reported success with all tests good!

**Comment by Cervator** [comment]
Tested out and merged - thanks! :-)

**Pull request merged**